### ISSUE TITLE
Failed to bundle into a single runnable bundle

### ISSUE THREAD

**Issue by grimly** [vercel/ncc#575]
Hello,

I've just discovered this tool thanks to my curiosity on how github actions works.

I maintain a build tool which have similar ideas, being a bundler with zero configuration for web code. basically I run tsc and rollup + babel in sequence.
I tried to apply to this build tool. My hopes to use your bundler are as follow :
* No more thousands dependencies just for the build tools.
  I still have a huge dependency tree behind my tool, none of it is useful to my projets.
  This would make my node_module folders have hundreds of files at most in my projets instead of tens of thousands
* If the "v8 compile cache" is what I think, it would cut my build time by half.

**Now my issue is as follow**

I've made it simple stupid with only my tsc task for a start.
I have a `src/main.ts` file as follow (code inspired from [typescript wiki](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API)) :

```typescript
import { createProgram, CompilerOptions, ScriptTarget, ModuleKind, getPreEmitDiagnostics, Diagnostic, flattenDiagnosticMessageText } from "typescript";

function printDiagnostics(diagnostics: ReadonlyArray<Diagnostic>) {
    diagnostics.forEach(diagnostic => {
        if (diagnostic.file) {
            let { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!);
            let message = flattenDiagnosticMessageText(diagnostic.messageText, "\n");
            console.log(`${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`);
        } else {
            console.log(flattenDiagnosticMessageText(diagnostic.messageText, "\n"));
        }
    });
}

const file = 'index.ts';
const options: CompilerOptions = {
    module: ModuleKind.ESNext,
    noEmitOnError: true,
    strict: true,
    target: ScriptTarget.Latest,
};
const program = createProgram([ file ], options);
const preDiagnostics = getPreEmitDiagnostics(program);
printDiagnostics(preDiagnostics);
const emitResult = program.emit();
printDiagnostics(emitResult.diagnostics);
```

**Time to build** : `yarn build` with in the package.json `"build": "ncc build src/main.ts -m --v8-cache"`.

**Time to try**
I have a `test/index.ts` file with this content :

```typescript
export function hello() {
    return "Hello world !"
}
```

Now I run it with this command : `pushd test; node ../dist; popd`

I get the following error :
```
File 'D:/Developments/Projets/typescript/workspaces/packagr/dist/lib.esnext.full.d.ts' not found.
```
(the mix of windows path and unix-like command is because I use gitbash)

Now this is embarrassing because the file exists but in a subfolder :
```
$ ls -l D:/Developments/Projets/typescript/workspaces/packagr/dist/typescript/lib/lib.esnext.full.d.ts
-rw-r--r-- 1 Unas 197121 1038 Aug 13 23:46 D:/Developments/Projets/typescript/workspaces/packagr/dist/typescript/lib/lib.esnext.full.d.ts
```

Could you point me what I did wrong ?

**Comment by grimly** [comment]
OK, I've found why it's going wrong.

The `typescript.js` file is bundled into the output `index.js` while other files are not. And unfortunately, lib files are not in the same folder.

To fix it, I have to provide a modified host as such :

```typescript
const defaultHost = createCompilerHost(options);
const host: CompilerHost & Required<Pick<CompilerHost, 'getDefaultLibLocation'>> = {
    ...defaultHost,
    getDefaultLibLocation() {
        return path.join(__dirname, 'typescript', 'lib');
    },
    getDefaultLibFileName(opts) {
        return path.join(this.getDefaultLibLocation(), getDefaultLibFileName(opts));
    }
};
```

I'm unsure it will solve everything because an other `__filename` is located in the `sys` namespace and I don't know yet if it is critical (it looks like it is used for localization but I don't know how to verify it)


Basically, ncc may break code relying on `__filename` or `__dirname`.
